### ISSUE TITLE
Consider a user preference for "Advanced" link editing

### ISSUE THREAD

**Issue by richtabor** [WordPress/gutenberg#60444]
I'd like to discuss considering a preference for how one creates and edits links throughout the editor. 

Based on the feedback on https://github.com/WordPress/gutenberg/issues/50891 it's clear that no one solution works for everyone. 

### Current, with link preview
When you select, or insert, a link, the link's "preview" is rendered, which presents the external link, URL, and edit/unlink/copy controls. If you want to edit the link further—like, open it in a new tab—you invoke the "edit" view, which has additional link controls. 

![CleanShot 2024-04-03 at 16 42 52](https://github.com/WordPress/gutenberg/assets/1813435/0caa91bc-8506-46d2-932d-0d229bbe4128)

### Advanced, without link preview

https://github.com/WordPress/gutenberg/pull/50998 explores removing the "preview" from Link Control, where when you click on a link, you see the full "edit" view of the link instead—there's no preview.  **This could become the "Advanced" link editing experience**, where when you create (or click) a link, you skip the link preview step completely. 

![CleanShot 2024-04-03 at 16 43 41](https://github.com/WordPress/gutenberg/assets/1813435/813c1fcb-c154-407c-94e6-e142e048ca6d)

### Alternatives

This "Advanced" flow becomes _the_ link editing experience. 

I'm not sold on this, as every time you add a link, you'll see the edit view of a link, which has a substantially larger footprint. And if we add unlink and copy link controls (which are likely necessary), the view grows in complexity even further. 

Thoughts? @WordPress/outreach

If we think this is a good approach, I don't love "Advanced" link editing—let's think of potential alternatives.

**Comment by unscripted** [comment]
I've ran into a few sites recently where this would be really helpful. Especially when working on sites with heavy outbound link that open in a new tab. 

For clarification, when you say "user preference," would this be a setting a user could set similar to always showing the list view?

What about Link Details or Link Options instead of Advanced?

**Comment by markhowellsmead** [comment]
The process to add a link using the modal and then set it to open in a new window requires multiple clicks, which can be inconvenient and more time-consuming for users who are adding many links to a long text in this way. It would be more convenient if the user didn't have to “hunt“ for the option to open in a new window. I've received this feedback from a couple of clients before. 

It would also make sense that the linking workflow is the same in all instances, be it on a text within a paragraph, on an image, or anywhere else.

**Comment by firepainting** [comment]
@richtabor 

>Based on the feedback on #50891 it's clear that no one solution works for everyone.

After commenting on the other thread, it's now clearer to me what the problem here is: You guys thought there was a "problem" that needed a "solution", when in fact, at least to me, it wasn't a problem, if at all!

>I'm not sold on this, as every time you add a link, you'll see the edit view of a link, which has a substantially larger footprint. And if we add unlink and copy link controls (which are likely necessary), the view grows in complexity even further.

I wouldn't mind a larger footprint or some additional buttons. **What I do mind is the crazy amount of clicks just to create/edit links!**

>If we think this is a good approach, I don't love "Advanced" link editing—let's think of potential alternatives.

This is why, for multiple times, I asked for some real-world data to back up your claim that that Advanced link editing is a PITA. And still after a long time no one shows us that. You know why? _Cos it was perfectly fine._

If it's just you (or your peers) that doesn't like it and thus decided to just roll it out and affect thousands of users, then we have a bigger problem here.

Yes, I write long-form. Yes, my stuff could use a lot of external links because, hey, scientific data.

@unscripted
>For clarification, when you say "user preference," would this be a setting a user could set similar to always showing the list view?

I agree with the above. Someone's (especially a WP dev) preference shouldn't affect thousands of users just like that. If it's highly debatable/crazy, make it something anyone can just set/show/hide. Problem solved. No need to push a narrative, no need to shove it down someone's throat just because you think it's better (it's not).

@markhowellsmead 
>The process to add a link using the modal and then set it to open in a new window requires multiple clicks, which can be inconvenient and more time-consuming for users who are adding many links to a long text in this way. It would be more convenient if the user didn't have to “hunt“ for the option to open in a new window. I've received this feedback from a couple of clients before.

THANK YOU for affirming our sentiments.

**Comment by richtabor** [comment]
> I agree with the above. Someone's (especially a WP dev) preference shouldn't affect thousands of users just like that. If it's highly debatable/crazy, make it something anyone can just set/show/hide. Problem solved. No need to push a narrative, no need to shove it down someone's throat just because you think it's better (it's not).

You are misunderstanding.

A user preference that _you_ decide on. 

You either want the additional link options, or not. Perhaps it's a setting in the Preferences panel: 

![CleanShot 2024-04-04 at 13 51 23](https://github.com/WordPress/gutenberg/assets/1813435/8e663d1b-6a5a-4da0-8aa6-2ad258cf4700)

> If it's highly debatable/crazy, make it something anyone can just set/show/hide. Problem solved. No need to push a narrative, no need to shove it down someone's throat just because you think it's better (it's not).

That's precisely what I'm proposing.

**Comment by TapiwaZvaks** [comment]
@richtabor I don't know why you would want to hide this option that's possibly used by tens of thousands of people daily. But hey, you are the developer. We are mere users. If it brings you peace and if it brings closure to this year(s) old saga, that's fine. This is not the first time that you have been challenged to bring just one person who thought the experience before 6.3 was bad. Just one. 

You express worry about a large footprint. Was that the reason for bringing the link preview and hiding other useful stuff? If yes, you don't have to worry about that. Surely you know that nobody is trying to write blog posts from a mobile phone or even from a tablet, right? Our screens are big enough for the additional footprint. 

In summary, the link preview is pretty useless. What we need is to be immediately presented with the _open in new tab_ and _mark and no follow_ options. But based on the fact that we have been debating this for a year and you are still stuck in denial, I don't hold any hope. Personally, I am rolling back to 6.2 because the many unnecessary clicks have become rather maddening.

**Comment by markhowellsmead** [comment]
I don’t know the background to the decisions which have been made to bring the interface to the stage that it’s at currently. I think that the best way ahead is to gain a range of opinions and continue to work together to find the best way ahead. 

As a user, and as a recipient of feedback from other users and our own clients, user preferences are too hidden to be found easily, and usually require one-on-one training to surface. There seem to be more positives than negatives to a small increase in modal size, to allow regularly-used options to be visible. I don’t personally class “open in new window” to be an advanced function, rather a commonly-used one.

**Comment by bacoords** [comment]
From my understanding, one (not the only) motivation for this rebuild was making sure that the user flow for adding and editing links was fully accessible- which is always tricky with popups and modals that by nature interrupt inline text editing. This feature has gone through a number of iterations and has made a lot of progress- kudos to the developers here.

Re: Advanced Settings
I'm not a fan of adding additional user preferences in some settings screen- especially for something as trivial as this. What is the reason for even having an advanced tab? Can we not just show those two checkboxes at all times and eliminate the additional click? Heck you could even use Flex, put them inline and save the screen space (for English at least)

**Comment by richtabor** [comment]
> I'm not a fan of adding additional user preferences in some settings screen- especially for something as trivial as this. What is the reason for even having an advanced tab? Can we not just show those two checkboxes at all times and eliminate the additional click? Heck you could even use Flex, put them inline and save the screen space (for English at least)

@bacoords I'm referring to a user preference to omit the preview entirely, not the toggle (worth a separate discussion perhaps though).

**Comment by flexseth** [comment]
> In summary, the link preview is pretty useless.

In general, linking (to include "embedded" links - that largely don't work) in WordPress could use improvement.

It's going to be difficult to get broad adoption behind cases that reflect only a small population of what WordPress users need. Most WordPress users need a hyperlink to go somewhere, that's it. 

That said, if someone creates some type of Advanced Linking functionality (via a block plugin) - I could see huge support behind that and people building on top of it. 

I agree that links could be better. Rich previews would be great. Better UI? Sure

But I don't see that being supported as something that is required, as overhead, for the hundreds of millions of WordPress users out there. A community plugin could be perfectly suited here!

**Comment by andreawetzel** [comment]
As someone who works on a large multisite with site administrators of varying technical backgrounds, I really like the clean UI link redesign in 6.5 with the extra step to "open in new tab". 

The user preference toggle described here seems to be a good option for folks who want to see all of the options all of the time -- almost like enabling a "developer mode" for editing links. I also think a site-wide filter or community plugin would be good ideas to enable seeing the advanced tab automatically if folks want to implement it site-wide.

**Comment by flexseth** [comment]
> What I do mind is the crazy amount of clicks just to create/edit links!

`Cmd + k` or `Ctrl + k` with text selected is a shortcut for turning highlighted text text into a hyperlink in just about all modern software, including WordPress, Email and GitHub 

Click the link to change it's behavior or URL

**Comment by TapiwaZvaks** [comment]
@flexseth I really don't understand what you're saying. The argument here is that in 6.4 one needed 2 clicks to set a link to open in a new tab. In 6.5 one needs 5 clicks. Moreover this is a commonly used feature for writers. As evidenced by the hundreds, if not thousands of complaints all over the web. Its as simple as that. I don't know what you are saying about ctrl +k etc. @andreawetzel there are plugins for opening links in a new page. Again, thats not the point. How many extra plugins are we going to need before the guys finish decimating the interface? As to the Advanced option, why hide a feature thats used a tens of thousands of times daily. This is an essential feature. To be honest I have never used anything that currently within the advanced section. But if it would bring an end to this, by any means hide away

**Comment by flexseth** [comment]
> As to the Advanced option, why hide a feature thats used a tens of thousands of times daily.

The argument boils down to _What Should Core Do?_
It's a healthy conversation to be had... this is Open Source software. 
Comes up a couple of times every few years it seems...

**A better question:** How can Core be extended to do what _you_ want it to do? 

Quite easily. 

Create a `block hook` (filter) to add `BlockControls` when a `link` is _selected_ inside of a `block`


## Link tools
Someone could create an entire suite of tools for working with links in this way, rich embeds, etc..

It has been a long time coming. Chatter has been going on for a LONG time about how bad links are in WordPress (creating actual, good nav menus were a critical first step)

## Extending links example
Maybe a link that when, if you copy/paste longitude / latitude it displays a map. 
Or even a `block plugin` that creates maps when longitude/latitude are found in a post.. 

Almost 100% could guarantee it wouldn't be a good candidate for Core functionality, but helpful. 

But someone could build that and contribute it to the community, and it would probably be widely used.

Context: I came into the conversation without reading what was said and these are my opinions from a long time working with WordPress. It looks like I actually disagree with @richtabor on this one and he's amazing at UX.

**Comment by flexseth** [comment]
> The argument here is that in 6.4 one needed 2 clicks to set a link to open in a new tab. In 6.5 one needs 5 clicks.

Most sites don't need this behavior for links. If it's something that's needed a lot, a contributor supported plugin for modifying the behavior of links could suit the need?

I couldn't tell you the last time I've build a WordPress website and needed to worry about how many clicks it takes to open a link in a new window...

**Comment by flexseth** [comment]
> Moreover this is a commonly used feature for writers.

Here's a solution. The command palette offers the way to do stuff. Have someone come up with custom commands for the command palette that bring up the interface with the options you want :)

I'll be working with the command palette later this month and can possibly take a look. 

Understand the frustration with linking, it's been a pain point for a long time. I'm not trying to discount what you're saying at all.. just trying to provide some info from what I've seen as the editor has changed

10/4 over and out :)

**Comment by TapiwaZvaks** [comment]
@flexseth I suspected/suspect you were having an entirely different argument. I speak from the point of view of a writer, though I am able to cobble together decent WordPress websites as well as anyone. Primarily I am a writer. And I repeat its hair pulling frustrating to have to click 5 times where a week ago I had to click only 2 times. Moreover hiding useful stuff started in 6.3. We had an entire debate back then regarding the issue with developers. Work was done with the result that in 6.4, six clicks were reduced to only two once again. Unfortunately in 6.5, we are back to 5 clicks. I assume from your language that you are a developer. I am not. I am a user who needs to set at least 15 links in different articles to open in a new tab every day. Its unbelievably annoying to have to do 5 clicks for each. Just imagine the situation for someone working for a large newspaper. The major problem is that developers don't want to hear anything from users. They are determined to cleanup the interface. But WordPress is not some pet that needs to be beautified. After all our target audiences never see the backend. Instead the philosophy behind WordPress should be usability. Let's work on speed. Google is currently decimating blogs. What can we do to improve the situation? How can we improve tables so we can have less not more plugins? It seems to me the current philosophy is entirely misguided. If there are developers who are also serious bloggers they would confirm what we are saying. Unfortunately there is this dogged dogmatism. Developers control the canvas and they are determined to clean it up. Never mind the pain being foisted on actual users along the way.

**Comment by flexseth** [comment]
@TapiwaZvaks can you link to a sample article? I may can write a block plugin that provides extra options for links... however- I'm not the best developer so it might take a couple of weeks. I do sympathize from a content standpoint, used to run a site that had a lot of content and we used the [PressThis ](https://codex.wordpress.org/Press_This) bookmarklet a lot (somewhat deprecated, since)

Opening links in new windows is an accessibility concern, that may be why the interface was changed. There's been a big push for accessibility in the block editor recently
> If a link opens a new tab or window without warning, people who are blind or visually impaired and use a screen reader to navigate will likely not realize that a new window or tab has opened and, as a result, will not know where they are and how to get back to the previous page.

via [Equalize Digital](https://equalizedigital.com/accessibility-checker/link-opens-new-window-or-tab/#:~:text=If%20a%20link%20opens%20a,back%20to%20the%20previous%20page.)

I've been working with documentation recently and can see plenty of reasons to want to open a link in a new tab. Not against the idea, just suggesting that UI changes like this are never easy

Regarding the number of clicks being changed or the UI - honestly I haven't messed with it.
Sorry this is a problem on the sites you manage. WordPress is supposed to be easy..

Link me to a post if you want, and if you can provide the block editor content, I can try to write a plugin to look for links and provide more options?

**Comment by flexseth** [comment]
@TapiwaZvaks - looking back at this, I think the best way to do it would be via the Command Palette.

Highlight text, `Cmd + Shift + n` when text is highlighted brings up the link editor interface, with "open in a new window" checked by default.

 Everything else works as normal

This will have to be coded, but if it's something you've seen as a pain point, worth looking at!

**Comment by TapiwaZvaks** [comment]
Anyway the fact of wanting to be politically correct has also been brought up previously. Lets not treat visually impaired people as if they are witless please. In fact they tend to be more computer savvy than the average user. I went to college with a number of visually impaired friends. They put the rest of us to shame as far as navigating around a computer screen was concerned. Its not surprising though, considering how we live in a world where people try to be politically correct even about mundane things. I doubt visually impaired people have actually complained about new tabs.

 For me as a blogger, its beneficial to have people stay on my site as long as is possible. It makes no sense to have a person redirected to another website after reading only a couple of sentences on my site. Regarding the command palette, I am not technical. In 6.4 I used ctrl + v to paste a link url. That brought up the option to open in new tab. I only needed to click once more and my choice was automatically saved. 

Before making changes please consider usability. Its why plug-ins for opening in new tabs have hundreds of thousands of installs. Unfortunately I can't use a plug that uses Javascript on my amp sites.

 This is a simple enough feature that was there in the many years that I have been using WordPress. Its not a new invention. I don't see why its so complicated. 

Regarding the request for a sample page, well, any page on any WordPress website will do...I think

**Comment by markhowellsmead** [comment]
Discussion around the issue is welcome and helpful, however I think there may be some misunderstanding about "clicking" here. I get that there is frustration around the subject, but the problem addressed at the start of this Github issue will be better addressed and more easily solved if discussion remains focused. 

The topic of this discussion is around the fact that when adding a link to a piece of text—or to an image etc.—in the block editor, the user has to open the modal and add the link, then the modal closes automatically. The user then has to open the modal a second time and access the inline "advanced" settings to activate the checkbox "open in new window".  This means additional steps for all users, and may well create difficulties for users who have additional accessibility needs. 

This isn't something which an individual developer should need to build themselves. The UX needs to be improved.

Using Rich's original terminology at the beginning of this Github issue, I feel that the "preview" mode should only be presented when clicking on a link _which has previously been set_, so that the user can see what the link is, and then optionally click on one of the choices presented in the screenshot at the top of this issue.

When initially adding the link, the "edit" view should remain open until it's defocused or dismissed.

Using keyboard shortcuts is a valid suggestion, but these should be viewed as additional, optional ways to access the "edit" view. I've submitted this as a new topic for discussion at https://github.com/WordPress/gutenberg/issues/60498.

**Comment by firepainting** [comment]
@richtabor 

>A user preference that you decide on.

Gotcha.

Just so everyone is on the same page, let me provide a TL;DR of the stuff that transpired. You can [read how the drama unfolded here](https://github.com/WordPress/gutenberg/issues/50891#issuecomment-1682012915) (around August 2023). Also I'm just a WP normie, not a dev.

- After a WP update, the tick-box for "Open in new tab" gets tucked behind the "Advanced" part of the dialog box for creating or editing a link. Before, upon Ctrl-K the tick-box was there right away. But then it turned to some 3 or 4 more clicks. To some of us, this was definitely a horrible downgrade in UX. Hiding that tick-box made absolutely no sense. 
- Around that time forum posts on wordpress.org seeking help and clarification about the issue were closed and peeps were left to fend for themselves. #50891 supposedly addressed the problem (I think??), but that is now closed.
- We ask the devs what made them tuck "Open in new tab" behind more clicks. They can't give a valid reason. (Again [check out the long discussions here](https://github.com/WordPress/gutenberg/issues/50891#issuecomment-1682012915).)
- Our reason why we want the tick-box: Because we are SEO bloggers. We don't want visitors leaving our site. We publish lots. Besides, WP is already mature--it's about time it did not rely on a third-party plugin just to "Open in new tab" (or even NoFollow).
- We wait. Another WP update and issue still isn't resolved. A workaround is found (Ctrl-V). We still don't get the logic behind such move (hiding the tick-box).
- WP 6.5 rolls out, now "Open in new tab" is 5 or so clicks more.
- This Github issue opens because the discussion over at the now-closed #50891 continues.

That's it. Really it's just "Open in new tab" as far as I know. But a lot of noise has buried it. 

Now this particular thread itself "Consider a user preference for "Advanced" link editing" isn't really something I care about--**I just want the tick-box "Open in new tab" back when I hit Ctrl-K**. It did "improve" to about 2 clicks, but then it's become a circus come the 6.5 update. Lotsa noise.

@markhowellsmead 
>I think that the best way ahead is to gain a range of opinions and continue to work together to find the best way ahead.

That's what we've been telling the WP guys. Before rolling out some changes, at least run some poll or something. Seeing "Open in new tab" upon Ctrl+K was perfectly fine. Literally no one--except these WP devs--has complained because of it.

>user preferences are too hidden to be found easily, and usually require one-on-one training to surface.

IMO this discussion about user preference is just another noise product. They don't address our specific issue and now they "find" this user preference "problem". As someone said, it feels like these WP devs are just trolling us at this point.

>I don’t personally class “open in new window” to be an advanced function, rather a commonly-used one.

EXACTLY.

@bacoords 
>Can we not just show those two checkboxes at all times and eliminate the additional click?

The question here is why did they hide those checkboxes in the first place? Was it really horrible UI/UX?

@flexseth
>Most sites don't need this behavior for links. If it's something that's needed a lot, a contributor supported plugin for modifying the behavior of links could suit the need?
>I couldn't tell you the last time I've build a WordPress website and needed to worry about how many clicks it takes to open a link in a new window...

The thought process here is: "Open in new tab" was just right there, and now we're talking about another plugin because they hid it?

If you write and publish a lot and suddenly "Open in new tab" now needs 5 clicks, you'd understand...

@markhowellsmead
>The topic of this discussion is around the fact that when adding a link to a piece of text—or to an image etc.—in the block editor, the user has to open the modal and add the link, then the modal closes automatically. The user then has to open the modal a second time and access the inline "advanced" settings to activate the checkbox "open in new window". This means additional steps for all users, and may well create difficulties for users who have additional accessibility needs.
>I've submitted this as a new topic for discussion at...

Thank you. Again, as a normie, my (and others') issue is particularly on the ridiculously many clicks to "Open in new tab". It's been tiresome but I'm glad this issue's still moving forward.

**Comment by markhowellsmead** [comment]
Thanks for taking the time to bring use newbies (to the issue) up to date, @firepainting. I'm both an editorial power-user and a designer/dev, but I'd never come across Ctrl + K in WordPress. :D

I can see that there is frustration here and lack of understanding/explanation about why the change (or UX “degradation” in many peoples eyes) was made in the first place. Everyone–myself included—implements features which need to be re-visited and improved later on: that's the nature of development. I feel that the best approach is to continue to iterate on the usability issues based on the considerable feedback available, rather than re-hashing a detrimental and somewhat pointless “them and us” conversation.
